File: pypi/trackfw/serve/api_board.py
"""
serve/api_board.py — Board API: retorna roadmaps agrupados por estado (kanban).
Espelho Python de internal/serve/api_board.go e npm/src/serve/api_board.js.
"""

import os

from trackfw import config as _config
from trackfw.pathfmt import normalize_ref_separator

STATES = ["wip", "backlog", "blocked", "done", "abandoned"]


def _extract_title(content, filename):
    """Extrai título da primeira linha '# ...' ou usa o nome do arquivo."""
    for line in content.split("\n"):
        stripped = line.strip()
        if stripped.startswith("# "):
            return stripped[2:].strip()
    # fallback: nome sem extensão
    return os.path.splitext(filename)[0]


def _scan_state_dir(dir_path, state, agent=None):
    """Varre um diretório de estado e retorna lista de cards."""
    cards = []
    if not os.path.isdir(dir_path):
        return cards
    try:
        files = sorted(
            f for f in os.listdir(dir_path)
            if f.endswith(".md") and not os.path.isdir(os.path.join(dir_path, f))
        )
    except OSError:
        return cards

    for filename in files:
        full_path = os.path.join(dir_path, filename)
        content = ""
        try:
            with open(full_path, "r", encoding="utf-8") as f:
                content = f.read()
        except OSError:
            pass

        title = _extract_title(content, filename)
        # path relativo para uso pelo frontend — IDENTIFICADOR emitido em JSON
        # (ADR-2026-09-04, D1 categoria 2). Antes do ML-2A era um .replace() inline;
        # passou ao ponto único do runtime (D3). Comportamento idêntico.
        #
        # 🔴 full_path permanece nativo e é o que vai a open() (ADR D2).
        rel_path = normalize_ref_separator(os.path.relpath(full_path, os.getcwd()))

        card = {
            "file": filename,
            "title": title,
            "state": state,
            "agent": agent or "",
            "path": rel_path,
        }
        cards.append(card)

    return cards


def get_board(cfg):
    """
    Retorna dict com 'columns' (por estado) e 'agents' detectados.
    Suporta namespacing flat e by_agent.
    """
    roadmap_dir = cfg.get("roadmap_dir", "docs/roadmaps")
    namespacing = cfg.get("roadmap_namespacing", "flat")

    columns = {state: [] for state in STATES}
    agents_found = set()

    if namespacing == "by_agent":
        agents = _config.resolve_agent_namespaces(cfg, roadmap_dir)

        for agent in agents:
            for state in STATES:
                dir_path = os.path.join(roadmap_dir, agent, state)
                cards = _scan_state_dir(dir_path, state, agent)
                if cards:
                    agents_found.add(agent)
                columns[state].extend(cards)
    else:
        for state in STATES:
            dir_path = os.path.join(roadmap_dir, state)
            cards = _scan_state_dir(dir_path, state)
            columns[state].extend(cards)

    return {
        "columns": columns,
        "agents": sorted(agents_found),
    }

File: npm/src/serve/api_board.js
'use strict'

const fs = require('fs')
const path = require('path')
const { resolveAgentNamespaces } = require('../validator')
const { normalizeRefSeparator } = require('../lib/pathfmt')

const STATES = ['wip', 'backlog', 'blocked', 'done', 'abandoned']

/**
 * extractTitle retorna o título do arquivo: primeira linha '# ...' ou nome do arquivo sem extensão.
 * @param {string} content
 * @param {string} filename
 * @returns {string}
 */
function extractTitle(content, filename) {
  for (const line of content.split('\n')) {
    const trimmed = line.trim()
    if (trimmed.startsWith('# ')) {
      return trimmed.slice(2).trim()
    }
  }
  return filename.replace(/\.md$/, '')
}

/**
 * scanState lê todos os .md de um diretório e retorna os itens do kanban.
 * @param {string} dir
 * @param {string} state
 * @param {string} agent - agente ou '' para flat
 * @param {string} roadmapDir - diretório base (para montar path relativo)
 * @returns {Array<{file: string, title: string, state: string, agent: string, path: string}>}
 */
function scanState(dir, state, agent, roadmapDir) {
  const items = []
  let files = []
  try {
    files = fs.readdirSync(dir).filter(f => {
      if (!f.endsWith('.md')) return false
      try { return !fs.statSync(path.join(dir, f)).isDirectory() } catch (_) { return false }
    })
  } catch (_) {
    return items
  }

  for (const file of files) {
    let content = ''
    try { content = fs.readFileSync(path.join(dir, file), 'utf8') } catch (_) {}
    const title = extractTitle(content, file)
    // IDENTIFICADOR emitido em JSON, nao caminho de travessia (ADR-2026-09-04, D1
    // categoria 2): o frontend devolve este valor verbatim em GET /api/file?path=...,
    // onde o servidor refaz path.join — no Windows o join reconverte "/" para o
    // separador nativo, entao o round-trip fecha. O path.join da linha 47 acima, que
    // e o caminho realmente lido por fs.readFileSync, permanece intocado: as duas
    // expressoes sao independentes sobre os mesmos operandos (ADR D2).
    const relPath = normalizeRefSeparator(agent
      ? path.join(roadmapDir, agent, state, file)
      : path.join(roadmapDir, state, file))

    items.push({ file, title, state, agent: agent || '', path: relPath })
  }
  return items
}

/**
 * handleBoard responde ao GET /api/board com o kanban em JSON.
 * @param {object} cfg - configuração do trackfw
 * @param {http.IncomingMessage} req
 * @param {http.ServerResponse} res
 */
function handleBoard(cfg, req, res) {
  const roadmapDir = cfg.roadmapDir || 'docs/roadmaps'
  const namespacing = cfg.roadmapNamespacing || 'flat'

  const columns = { wip: [], backlog: [], blocked: [], done: [], abandoned: [] }
  const agentSet = new Set()

  if (namespacing === 'by_agent') {
    const agents = resolveAgentNamespaces(cfg, roadmapDir)
    for (const agent of agents) {
      agentSet.add(agent)
      for (const state of STATES) {
        const dir = path.join(roadmapDir, agent, state)
        const items = scanState(dir, state, agent, roadmapDir)
        columns[state].push(...items)
      }
    }
  } else {
    for (const state of STATES) {
      const dir = path.join(roadmapDir, state)
      const items = scanState(dir, state, '', roadmapDir)
      columns[state].push(...items)
    }
  }

  const result = {
    columns,
    agents: Array.from(agentSet).sort(),
  }

  res.writeHead(200, { 'Content-Type': 'application/json' })
  res.end(JSON.stringify(result))
}

module.exports = { handleBoard }
